junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,447 @@
+    List(ScalaJSJUnitPluginComponent)
+
+  val description: String = "Adapts JUnit test classes to Scala.js"
+
+  object ScalaJSJUnitPluginComponent extends plugins.PluginComponent
+      with transform.Transform {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = selfPlugin.global
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends global.Transformer {
+
+      import global._
+
+      type ScalaJSPlugin = NscPlugin {
+        def PrepInteropComponent: AnyRef {
+          def genExportMember(sym: Symbol): List[Tree]
+          def registerModuleExports(sym: Symbol): Unit
+          def genNamedExport(defSym: Symbol, jsName: String, pos: Position): Unit
+        }
+      }
+
+      lazy val scalaJSPlugin = {
+        global.plugins.collectFirst {
+          case pl if pl.getClass.getName == "org.scalajs.core.compiler.ScalaJSPlugin" => pl
+        }.get.asInstanceOf[ScalaJSPlugin]
+      }
+
+      override def transform(tree: Tree): Tree = tree match {
+        case tree: PackageDef =>
+          def classHasJUnitAnnotations(clazz: ClassDef): Boolean = {
+            clazz.impl.body.exists {
+              case ddef: DefDef => hasAnnotation[org.junit.Test](ddef)
+              case _ => false
+            }
+          }
+          val newStats = tree.stats.groupBy {
+            case clDef: ClassDef => clDef.name
+            case _ => "not a class def"
+          }.iterator.flatMap {
+            case ("not a class def", xs) => xs
+            case (className, x :: xs) =>
+              val clDef1 = x.asInstanceOf[ClassDef]
+              val clDef2Option = xs match {
+                case Nil => None
+                case y :: ys =>
+                  if (ys.isEmpty) Some(y.asInstanceOf[ClassDef])
+                  else None
+              }
+              val (clDefOption, mlDefOption) = {
+                if (clDef1.mods.hasFlag(Flags.MODULE))
+                  (clDef2Option, Some(clDef1))
+                else (Some(clDef1), clDef2Option)
+              }
+
+              if ((clDefOption.isDefined && classHasJUnitAnnotations(clDefOption.get)) ||
+                (mlDefOption.isDefined && classHasJUnitAnnotations(mlDefOption.get))) {
+                val transformedClDef = clDefOption match {
+                  case Some(clDef) => transformTestClass(clDef)
+
+                  case None =>
+                    throw new ClassNotFoundException(
+                      mlDefOption.get.name.toString)
+                }
+                val hookClass = mkHookClass(transformedClDef, mlDefOption)
+                hookClass :: transformedClDef :: mlDefOption.toList
+              }
+              else x :: xs
+
+            case (_, Nil) => Nil
+          }
```

Comment:
How about something like this?

```scala
val clDefs = tree.stats.collect {
  case clDef: ClassDef => clDef
}

val companionPairs = clDefs.groupBy(_.name).iterator

val newClasses = companionPairs.flatMap { (className, companions) =>
  val needsTransform = companions.exists(classHasJUnitAnnotations)

  if (needsTransform) {
    val (cls, mods) = companions.partition(_.symbol.isModuleClass)

    assert(cls.size <= 1, "More than one class with the same name")
    assert(mods.size <= 1, "More than one module with the same name")

    cls match {
      case List(clDef) =>
        val transformedClDef = transformTestClass(clDef)
        val hookClass = mkHookClass(transformedClDef, mods.headOption)
        hookClass :: transformedClDef :: mods
      case Nil =>
        // fail nicely (why do you look for annotations in the module
        // class if you don't expect them?)
      case _ =>
        throw new AssertionError("More than one class with the same name")
    }
  } else {
    companions
  }
}

val newStats = newClasses ++ tree.stats.filterNot(_.instanceOf[ClassDef])
```

Also, are we sure that reordering stats is always fine here? Optionally, you could patch the stats with the modified class trees where necessary to retain ordering.